File: store/auth/actions.ts
import { ImmerAction, ImmerState } from '@/store/types';
import { subState } from '@/store/data';
import { api } from '@/services/api';

import { AuthLoginPayload, AuthStore, User } from './types';
import * as utils from './utils';
import { authState } from './data';

export const login =
  (set: ImmerAction<AuthStore>) => async (payload: AuthLoginPayload) => {
    set((state) => {
      state.state.login = { ...subState, loading: true };
    });
    try {
      const res = await api.post<User>('/auth/v1/log_in', payload);
      const data = res.data;

      if (!res.data.isVerified || res.data.role === 'user') {
        utils.deleteCookie('id');
        throw new Error('User not allowed to access this page.');
      }

      set((state) => {
        state.state.login = { ...subState, success: true };
        state.data.user = data;
        // Update state.data.login here
      });
    } catch (error) {
      set((state) => {
        state.state.login = { ...subState, error: true };
      });
    }
  };

export const logout = (set: ImmerAction<AuthStore>) => async () => {
  set((state) => {
    state.state.logout = { ...subState, loading: true };
  });
  try {
    // Add your API call here
    set((state) => {
      state.state.logout = { ...subState, success: true };
      // Update state.data.logout here
    });
  } catch (error) {
    set((state) => {
      state.state.logout = { ...subState, error: true };
    });
  }
};

export const init = (set: ImmerAction<AuthStore>) => async () => {
  set((state) => {
    state.state.init = { ...subState, loading: true, init: true };
  });
  try {
    const res = await api.get<User>('/user/v1/get');
    if (!res.data.isVerified || res.data.role === 'user') {
      utils.deleteCookie('id');
      throw new Error('User not allowed to access this page.');
    }
    set((state) => {
      state.state.init = { ...subState, success: true, init: true };
      state.data.user = res.data;
      // Update state.data.init here
    });
  } catch (error) {
    set((state) => {
      state.state.init = {
        ...subState,
        error: true,
        init: true,
        message: error.message
      };
    });
  }
};

export const reset = (set: ImmerAction<AuthStore>) => async () => {
  set((state) => {
    state.state = { ...authState.state };
    state.data = { ...authState.data };
  });
};

File: admin-dioxus/src/store/auth/actions.rs
use super::{AuthState, AuthUser, LoginPayload, UserRole};
use crate::{services::http_client, store::StateFrame};
use dioxus::{logger::tracing, prelude::*};

impl AuthUser {
    pub fn new(id: i32, name: String, email: String, role: UserRole, is_verified: bool) -> Self {
        AuthUser {
            id,
            name,
            email,
            avatar: None,
            role,
            is_verified,
        }
    }

    pub fn dev() -> Self {
        AuthUser::new(
            1,
            "Dev User".to_string(),
            "dev@example.com".to_string(),
            UserRole::Admin,
            true,
        )
    }
}

impl AuthState {
    pub fn new() -> Self {
        AuthState {
            user: GlobalSignal::new(|| None),
            login_status: GlobalSignal::new(|| StateFrame::new()),
            logout_status: GlobalSignal::new(|| StateFrame::new()),
            init_status: GlobalSignal::new(|| StateFrame::new()),
        }
    }

    pub async fn logout(&self) {
        self.logout_status.write().set_loading(None);
        let empty_body = {};
        let result = http_client::post("/auth/v1/log_out", &empty_body)
            .send()
            .await;
        match result {
            Ok(response) => {
                if (200..300).contains(&response.status()) {
                    self.logout_status.write().set_success(None, None);
                    *self.user.write() = None;
                } else {
                    self.logout_status.write().set_api_error(&response).await;
                    *self.user.write() = None;
                }
            }
            Err(e) => {
                self.logout_status.write().set_failed(Some(e.to_string()));
                *self.user.write() = None;
            }
        }
    }

    pub async fn init(&self) {
        // self.init_status.write().set_success(None, None);
        // *self.user.write() = Some(User::dev());
        self.init_status.write().set_loading(None);
        let result = http_client::get("/user/v1/get").send().await;
        match result {
            Ok(response) => {
                if (200..300).contains(&response.status()) {
                    match response.json::<AuthUser>().await {
                        Ok(user) => {
                            if !user.is_verified || !user.is_admin() {
                                self.init_status.write().set_failed(Some(
                                    "User not allowed to access this page.".to_string(),
                                ));
                                return;
                            }
                            *self.user.write() = Some(user);
                            self.init_status.write().set_success(None, None);
                        }
                        Err(e) => {
                            tracing::error!("Failed to parse user data: {}", e);
                            self.init_status
                                .write()
                                .set_failed(Some(format!("Failed to parse user data: {}", e)));
                        }
                    }
                } else if response.status() == 401 {
                    // Unauthorized, no user logged in
                    self.init_status.write().set_success(None, None);
                } else {
                    self.init_status.write().set_api_error(&response).await;
                }
            }
            Err(e) => {
                self.init_status
                    .write()
                    .set_failed(Some(format!("Network error: {}", e)));
            }
        }
    }

    pub async fn login(&self, email: String, password: String) {
        self.login_status.write().set_loading(None);
        let payload = LoginPayload { email, password };
        let result = http_client::post("/auth/v1/log_in", &payload).send().await;
        match result {
            Ok(response) => {
                if (200..300).contains(&response.status()) {
                    match response.json::<AuthUser>().await {
                        Ok(user) => {
                            if !user.is_verified || !user.is_admin() {
                                self.login_status.write().set_failed(Some(
                                    "User not allowed to access this page.".to_string(),
                                ));
                                return;
                            }
                            *self.user.write() = Some(user);
                            self.login_status.write().set_success(None, None);
                        }
                        Err(e) => {
                            eprintln!("Failed to parse user data: {}", e);
                            self.login_status
                                .write()
                                .set_failed(Some(format!("Failed to parse user data: {}", e)));
                        }
                    }
                } else {
                    self.login_status.write().set_api_error(&response).await;
                }
            }
            Err(e) => {
                self.login_status
                    .write()
                    .set_failed(Some(format!("Network error: {}", e)));
            }
        }
    }

    pub fn reset(&self) {
        *self.user.write() = None;
        *self.login_status.write() = StateFrame::new();
        *self.logout_status.write() = StateFrame::new();
        *self.init_status.write() = StateFrame::new();
    }
}
